### DISCUSSION TITLE
Hard crash when applying suggested tsconfig.json in read-only environments

### DISCUSSION THREAD

**Discussion by tgolsson** [vercel/next.js#64638] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```
    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Wed Jun 16 23:47:55 UTC 2021
    Binaries:
      Node: 16.13.0
      npm: 8.1.0
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.0.11-canary.21
      react: 17.0.2
      react-dom: 17.0.2
```

### What browser are you using? (if relevant)

N/A

### How are you deploying your application? (if relevant)

N/A

### Describe the Bug

In some build environments (CI, Bazel, ...) the source code may be mounted at a read-only path. If next attempts to modify the tsconfig.json file it'll crash out the build with an error such as:

```
[Error: EACCES: permission denied, open '<snip>/test-next/nextjs-blog/tsconfig.json'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '<snip>/test-next/nextjs-blog/tsconfig.json'
}
```

### Expected Behavior

Not crashing. I'm happy that Next has opinions and helpful onboarding, but the error message has too little info and isn't actionable. Also, I'd expect the process to continue as long as it's not a user request to create the file -- I'd happily have a long error log due to missing an option, and find a warning about tsconfig.json at the top than a random undiagnosable crash. 

### To Reproduce

1. Follow the get started tutorial up until you're able to `$ npm run dev`.
2. `$ touch pages/foobar.ts`
3. `$ npm install --save-dev typescript @types/react @types/node`
4. `$ npm run build` so next.js generates a `tsconfig.json`
5. Edit `tsconfig.json` and remove a compiler option -- I assume all trigger it, but I removed the `moduleResolution` option.
6. `$ chmod -w tsconfig.json`
7. `$ npm run build`

Crash here: https://github.com/vercel/next.js/blob/canary/packages/next/lib/typescript/writeConfigurationDefaults.ts#L174. I assume it'll also crash here if it's first-time setup in a read-only environment: https://github.com/vercel/next.js/blob/canary/packages/next/lib/typescript/writeConfigurationDefaults.ts#L98

**Comment by balazsorban44** [comment]
Hi, thanks for reporting.

Just to understand your use case, in which scenario would you like to run the dev server in a CI environment?

Otherwise, I can see that `npm run build` will also fail as it also tries to touch that file, but I think it should be safe to assume that the `tsconfig.json` file should have been correctly configured before it reaches CI.

Next.js sets/overrides certain options for the most optimal output, and I can see no immediate changes to fix your problem, other than pushing a commit with the updated `tsconfig.json` file.

We could make the error message nicer, but in reality, it really should have been caught already in the development process.

Do you have a proposal for a fix/change?

**Comment by tgolsson** [comment]
> Just to understand your use case, in which scenario would you like to run the dev server in a CI environment?

Changed repro steps to `build`, the command is irrelevant for the problem at hand -- the description mentions crashing builds, and how that build occurs is irrelevant. I used `npm run dev` because I followed the get-started guide while creating the repro. We obviously do not run the development server in CI.

> Otherwise, I can see that npm run build will also fail as it also tries to touch that file, but I think it should be safe to assume that the tsconfig.json file should have been correctly configured before it reaches CI.

That's the point. It has been configured like we want. ;-) 

> We could make the error message nicer, but in reality, it really should have been caught already in the development process.

It boggles my mind to have a bug report to where "your tool crashes without diagnostic when it tries to change my build configuration without asking" and come back and say, "yeah that's your fault". I'm not a JS person but as far as I can tell we have a shared tsconfig.json between different projects, and in this case we had an override tsconfig.json for a specific project which got updated (and committed). So the error got thrown for a number of **other** projects in our repository in CI. 

> Do you have a proposal for a fix/change?

Yes. Do not crash (but log!), and if you have to crash, at least have the decency to tell me why and a fix. It's right there in my expectations. Literally, the below would be so much more helpful:

```js
try {
   // existing code
} catch(error) {
    something.error("We tried to change your build settings but did not succeeded. We suggest the following changes:");
    for (let x of suggestions) {
        something.println(`\t ${x.setting} => ${x.value}`);
    }

    something.println("Continuing with existing configuration.");
    return; 
}
```

Either it's suggested changes and you can progress, or it's a diagnosable build error and you log and exit.

**Comment by balazsorban44** [comment]
Thank you for the extra info!

Here is the reasoning of why Next.js does this currently:

> Next.js sets/overrides certain options for the most optimal output

I'll discuss with the others if we should either allow continuing the build with a warning that tells the user the output might not be optimal, or we should still crash but add a more clear error message with a link to a page that gives an actionable way to fix it.
